libraries/chain/controller.cpp

```diff
@@ -1877,20 +1877,31 @@ struct controller_impl {
 
       try {
 
+      uint32_t hs_lib = hs_irreversible_block_num.load();
+      const bool hs_active = hs_lib > 0;
+
       auto& pbhs = pending->get_pending_block_header_state();
 
       auto& bb = std::get<building_block>(pending->_block_stage);
 
       auto action_merkle_fut = post_async_task( thread_pool.get_executor(),
-                                                [ids{std::move( bb._action_receipt_digests )}]() mutable {
-                                                   return canonical_merkle( std::move( ids ) );
+                                                [ids{std::move( bb._action_receipt_digests )}, hs_active]() mutable {
+                                                   if (hs_active) {
+                                                      return calculate_merkle( std::move( ids ) );
+                                                   } else {
+                                                      return canonical_merkle( std::move( ids ) );
+                                                   }
```

Comment:
Can you make above code block line 1888 to 1893 as a method taking `hs_active` (or if_active) as a parameter? It can be reused in several places.